### PR TITLE
Caching

### PR DESCRIPTION
**Pull request by tanveergill**
##### Checklist

- [x] Tested in playground or other setup
- [ ] Screenshot (Grafana) from playground added to PR for 15+ minute run
- [ ] Documentation is changed or added
- [ ] Tests and/or benchmarks are included
- [ ] Breaking changes

## Summary by CodeRabbit

- **New Features**
  - Introduced caching capabilities to the FlowControlService with new RPCs for cache management.
  - Added cache key and value fields to request and response messages for enhanced service functionality.
  - Implemented a new cache package with upsert and delete operations for distributed caching.
  - Expanded SDKs with new cache-related types and methods for managing cached values.

- **Enhancements**
  - Updated SDKs to include new cache-related RPCs and message types.
  - Improved error handling and added new utility functions in SDKs for better integration with caching features.
  - Refactored existing code to accommodate caching logic and improve readability.

- **Documentation**
  - Updated example usage in SDKs to demonstrate new caching features and API changes.

- **Refactor**
  - Renamed packages and updated import paths to align with new caching functionality.
  - Replaced logging packages for consistency across the codebase.

- **Style**
  - Made minor adjustments to code formatting for better consistency and readability.

- **Chores**
  - Updated Makefile to copy new generated SDK stubs for flow control to respective SDK directories.

### CODE REVIEW

File: pkg/policies/flowcontrol/engine.go [hunk 1]
```diff
@@ -118,47 +120,79 @@ func (e *Engine) ProcessRequest(ctx context.Context, requestContext iface.Reques
...
+				Value:        cachedBytes,
+				LookupResult: flowcontrolv1.CacheLookupResult_HIT,
+				ResponseCode: flowcontrolv1.CacheResponseCode_SUCCESS,
 			}
+			response.DecisionType = flowcontrolv1.CheckResponse_DECISION_TYPE_ACCEPTED
 			return
 		}
+		response.CachedValue = &flowcontrolv1.CachedValue{
+			LookupResult: flowcontrolv1.CacheLookupResult_MISS,
+			ResponseCode: flowcontrolv1.CacheResponseCode_ERROR,
```
Comment (tanveergill) [inline pkg/policies/flowcontrol/engine.go:187]: Emit error only of it not key not found

**Review by harjotgill — APPROVED** [review]

File: sdks/aperture-js/sdk/cache.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    error: Error | null;
+    code: string | null;
+    message: string | null;
+};
+
+export type DeleteCachedValueResponse = {
+    error: Error | null;
+    code: string | null;
+    message: string | null;
+};
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply sdks/aperture-js/sdk/cache.ts:19]: @coderabbitai,
`lookupResult` is supposed to capture whether the cache lookup was a hit or a miss. Please suggest a more appropriate name and data type for this field?

[the remark this answers is not in the document]
Comment (tanveergill) [reply sdks/aperture-js/sdk/cache.ts:19]: How about `lookupStatus` instead of `cacheStatus`?

[the remark this answers is not in the document]
Comment (tanveergill) [reply sdks/aperture-js/sdk/cache.ts:19]: Can I leverage the auto-generated types in the protobuf instead:

```// Original file: proto/flowcontrol/check/v1/check.proto

export const CacheLookupStatus = {
  HIT: 0,
  MISS: 1,
} as const;

export type CacheLookupStatus =
  | 'HIT'
  | 0
  | 'MISS'
  | 1

export type CacheLookupStatus__Output = typeof CacheLookupStatus[keyof typeof CacheLookupStatus]```

[the remark this answers is not in the document]
Comment (tanveergill) [reply sdks/aperture-js/sdk/cache.ts:19]: Code is captures Error or Success of cache lookup. Can you suggest a better name?

File: sdks/aperture-js/sdk/cache.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  GetError(): Error | null {
+    return this.error;
+  }
+
+  GetOperationStatus(): string | null {
+    return this.operationStatus;
+  }
+}
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply sdks/aperture-js/sdk/cache.ts:100]: good catch!

File: sdks/aperture-go/sdk/flow.go [hunk 4]
```diff
@@ -1,58 +1,68 @@
...
 
 // ShouldRun returns whether the Flow was allowed to run by Aperture Agent.
 // By default, fail-open behavior is enabled. Set rampMode to disable it.
 func (f *flow) ShouldRun() bool {
-	if (!f.rampMode && f.checkResponse == nil) || (f.checkResponse.DecisionType == checkproto.CheckResponse_DECISION_TYPE_ACCEPTED) {
+	if (!f.rampMode && f.checkResponse == nil) || (f.checkResponse.DecisionType == checkv1.CheckResponse_DECISION_TYPE_ACCEPTED) {
 		return true
 	} else {
 		return false
 	}
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply sdks/aperture-go/sdk/flow.go:61]: okay

File: sdks/aperture-js/sdk/flow.ts [hunk 5]
```diff
@@ -56,6 +72,95 @@ export class Flow {
...
+          }
+          const resp = new SetCachedValueResponse(
+            ConvertCacheError(res?.error),
+            ConvertCacheOperationStatus(res?.operationStatus),
+          );
+          resolve(resp);
+        },
+      );
+    });
+  }
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply sdks/aperture-js/sdk/flow.ts:105]: This API is such that it always resolves and puts the error as a field inside returned value

**Pull request merged**